daemon/api_system_volumes.go

```diff
@@ -21,19 +21,167 @@ package daemon
 
 import (
 	"encoding/json"
+	"fmt"
 	"net/http"
 
+	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/gadget/device"
 	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/devicestate"
 	"github.com/snapcore/snapd/overlord/fdestate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
 	"github.com/snapcore/snapd/secboot/keys"
+	"github.com/snapcore/snapd/strutil"
 )
 
 var systemVolumesCmd = &Command{
 	Path:        "/v2/system-volumes",
+	GET:         getSystemVolumes,
 	POST:        postSystemVolumesAction,
+	ReadAccess:  rootAccess{},
 	WriteAccess: rootAccess{},
 }
 
+var (
+	fdeMgrGetKeyslots         = (*fdestate.FDEManager).GetKeyslots
+	fdeMgrGenerateRecoveryKey = (*fdestate.FDEManager).GenerateRecoveryKey
+	fdeMgrCheckRecoveryKey    = (*fdestate.FDEManager).CheckRecoveryKey
+	snapstateGadgetInfo       = snapstate.GadgetInfo
+)
+
+type systemVolumesResponse struct {
+	ByContainerRole map[string]volumeInfo `json:"by-container-role,omitempty"`
+}
+
+type volumeInfo struct {
+	ContainerRole string
+	VolumeName    string                 `json:"volume-name"`
+	Name          string                 `json:"name"`
+	Encrypted     bool                   `json:"encrypted"`
+	Keyslots      map[string]keyslotInfo `json:"keyslots,omitempty"`
+}
+
+type keyslotInfo struct {
+	Type fdestate.KeyslotType `json:"type"`
+	// only for platform key slots
+	Roles        []string        `json:"roles,omitempty"`
+	PlatformName string          `json:"platform-name,omitempty"`
+	AuthMode     device.AuthMode `json:"auth-mode,omitempty"`
+}
+
+func getSystemVolumes(c *Command, r *http.Request, user *auth.UserState) Response {
+	query := r.URL.Query()
+	containerRoles := query["container-role"]
+	byContainerRole := query.Get("by-container-role") == "true"
+	if len(containerRoles) > 0 && byContainerRole {
+		return BadRequest(`query parameter "container-role" conflicts with "by-container-role"`)
+	}
+
+	volumes, err := getAllVolumes(c.d.overlord.State(), c.d.overlord.FDEManager())
+	if err != nil {
+		return InternalError(err.Error())
+	}
+
+	res := systemVolumesResponse{
+		ByContainerRole: make(map[string]volumeInfo),
+	}
+	for _, volume := range volumes {
+		switch {
+		case len(containerRoles) > 0:
+			if strutil.ListContains(containerRoles, volume.ContainerRole) {
+				res.ByContainerRole[volume.ContainerRole] = volume
+			}
+		case byContainerRole:
+			res.ByContainerRole[volume.ContainerRole] = volume
+		default:
+			// all groupings, currently only by-container-role is supported.
+			res.ByContainerRole[volume.ContainerRole] = volume
+		}
+	}
+
+	return SyncResponse(res)
+}
+
+func getAllVolumes(st *state.State, fdemgr *fdestate.FDEManager) ([]volumeInfo, error) {
```

Comment:
it feels like this should be a helper in devicestate using things maybe helper from fdestate